pkg/kgo/record_and_fetch.go

```diff
@@ -139,6 +146,25 @@ type Record struct {
 	Offset int64
 }
 
+// WithContext enriches the Record with a Context.
+func (r *Record) WithContext(ctx context.Context) *Record {
```

Comment:
What's the use case for WithContext? I think this would only be necessary for producing, which already currently accepts a context to both Produce and ProduceSync. It may be less confusing to set `r.ctx = ctx` inside of Produce, rather than having two areas for setting the context and people not knowing which one actually controls cancelling, etc.

Reply:
The use case is to enrich consumer records. In a consumer hook, we would extract trace data from the Kafka headers to a context and enrich the record with that context. This makes propagating trace data very convenient and save users the hassle of manual work. They would use the consumer record context to do further tracing in their consumer logic. 

The consumer hook implementation would look like this:

```
func (k *Kotel) OnFetchRecordBuffered(r *kgo.Record) {
	textMapPropagator := propagation.NewCompositeTextMapPropagator(propagation.TraceContext{})
	ctx := textMapPropagator.Extract(context.Background(), NewRecordCarrier(r))
	r = r.WithContext(ctx)
        ...
}
```

Understandably, having two areas for setting the context is confusing. We are open to any suggestions you see necessary.

Reply:
I see, I was hoping that if I initialized the context internally, there would not need to be any public setter.

Reply:
Maybe you could provide a mechanism for setting the Context, `kgo.WithContextSetter(func(r *kgo.Record){} ctx.Context)` or a hook with a context return value?

Reply:
I'm leaning to adding a public Context (or Ctx) field, rather than a getter / setter. If only read access were needed, a method would be better, but write access complicates things.

Reply:
@twmb Okay, that works, I will think it is important that the context et being set (copied over) in the `produce` call. Otherwise it will be impossible to access the context from the produce related hooks, and thus create and end spans.

Thanks for thinking about this, we understand that adding a field to the Record struct for the use case of OpenTelemetry is controversial, but I dont really think there is any other way if franz-go needs to support OpenTelemetry in a seamless manner.